Fixes #10609: Adds lint to detect construction of unit struct using `default`

Icxolu opened the pull request:
Using `default` to construct a unit struct increases code complexity and adds a function call. This can be avoided by simply removing the call to `default` and simply construct by name.

changelog: [`default_constructed_unit_structs`]: detects construction of unit structs using `default`

fixes #10609

Manishearth commented on clippy_lints/src/unitstruct_default_construction.rs:
```diff
@@ -0,0 +1,70 @@
+impl LateLintPass<'_> for UnitstructDefaultConstruction {
+    fn check_expr<'tcx>(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>) {
+        // make sure we have a call to `Default::default`
+        let hir::ExprKind::Call(fn_expr, &[]) = expr.kind else {return};
```
nit: I would use `if_chain` for this row of if/if let + return

Icxolu replied:
done

Manishearth commented on clippy_lints/src/unitstruct_default_construction.rs:
````diff
@@ -0,0 +1,70 @@
+    /// }
+    /// ```
+    #[clippy::version = "1.71.0"]
+    pub UNITSTRUCT_DEFAULT_CONSTRUCTION,
````
issue: Not a fan of this name.

@rust-lang/clippy naming thoughts?

`default_constructed_unit` might be fine.

(I don't like `unitstruct`, i think `unit` or `unit_struct` is fine)

Icxolu replied:
I went for `default_constructed_unit_struct` for now, to avoid confusion with the unit type `()`. But if a better name comes up, I'm happy to change it again.

Manishearth replied:
yeah, that's fine by me. I'll wait for others to chime in

PatchMixolydic commented on clippy_lints/src/default_constructed_unit_struct.rs:
```diff
@@ -0,0 +1,66 @@
+                    cx,
+                    DEFAULT_CONSTRUCTED_UNIT_STRUCT,
+                    qpath.last_segment_span(),
+                    "Use of `default` to create a unit struct.",
```
According to the [rustc development guide](https://rustc-dev-guide.rust-lang.org/diagnostics.html#diagnostic-output-style-guide), diagnostic messages should not have capitalization or punctuation:
```suggestion
                    "use of `default` to create a unit struct",
```

Manishearth commented:
Some of the tests fail.

Also, this suggestion from the tests seems incorrect:

```
     fn vals(_: Self) -> Self {
-        Self::default()
+        Self::()
     }
```

Icxolu commented:
I updated the span to fix the broken suggestion and updated the uitest output to reflect this as well as the updated diagnostics message (forget about that in the last commit)

Manishearth commented:
Tests still fail, since a bunch of the other tests use this syntax. You should allow this lint in those tests, or update the test code to not have this problem (if possible)

flip1995 commented on clippy_lints/src/default_constructed_unit_struct.rs:
````diff
@@ -0,0 +1,66 @@
+    /// }
+    /// ```
+    #[clippy::version = "1.71.0"]
+    pub DEFAULT_CONSTRUCTED_UNIT_STRUCT,
````
I don't want this to get lost in the comment thread, so I'll leave this comment again here:

Please change the lint name to plural:

```suggestion
    pub DEFAULT_CONSTRUCTED_UNIT_STRUCTS,
```

Manishearth commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit 48ae5a071bc32e5fb69669248ab103ecba7509ab has been approved by `Manishearth`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=Manishearth 48ae5a071bc32e5fb69669248ab103ecba7509ab -->
<!-- homu: {"type":"Approved","sha":"48ae5a071bc32e5fb69669248ab103ecba7509ab","approver":"Manishearth","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit 48ae5a071bc32e5fb69669248ab103ecba7509ab with merge f9c1d155b4c9cad224fe96aad486993dc123c9b6...
<!-- homu: {"type":"BuildStarted","head_sha":"48ae5a071bc32e5fb69669248ab103ecba7509ab","merge_sha":"f9c1d155b4c9cad224fe96aad486993dc123c9b6"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/4876810159/jobs/8700767691), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/4876810161/jobs/8700774906), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/4876810162/jobs/8700950388)
Approved by: Manishearth
Pushing f9c1d155b4c9cad224fe96aad486993dc123c9b6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Manishearth","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/4876810162/jobs/8700950388","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/4876810159/jobs/8700767691","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/4876810161/jobs/8700774906"},"merge_sha":"f9c1d155b4c9cad224fe96aad486993dc123c9b6"} -->
